storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/wal"
+	fsnotify "gopkg.in/fsnotify/fsnotify.v1"
+)
+
+var (
+	segmentRegex    = regexp.MustCompile(`\d{8}`)
+	checkpointRegex = regexp.MustCompile(`checkpoint.\d{6}`)
+)
+
+type WriteTo interface {
+	Append([]tsdb.RefSample) bool
+	StoreSeries([]tsdb.RefSeries)
+	TrimSeries()
+}
+
+// WALWatcher watches the TSDB WAL for a given StorageClient (remote write config).
+type WALWatcher struct {
+	writer  WriteTo
+	logger  log.Logger
+	walDir  string
+	watcher *fsnotify.Watcher
+
+	segment         *wal.Segment
```

Comment:
Not keen on the quasi-global state of the `segment` field in the WALWatcher; feels like this should be passed between functions as needed?  Its mainly just used in `watch()`.  That should make the lifecycle of the object clearer.